File: generated-code/example/kotlin/TcpReadWrite/src/main/kotlin/model/OneOf.kt
package model

import util.StreamUtil

/**
 * Oneof example
 */
abstract class OneOf {
    /**
     * Write OneOf to output stream
     */
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)

    companion object {
        /**
         * Read OneOf from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): OneOf {
            when (StreamUtil.readInt(stream)) {
                OptionOne.TAG -> return OptionOne.readFrom(stream)
                OptionTwo.TAG -> return OptionTwo.readFrom(stream)
                else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }

    /**
     * First option
     */
    class OptionOne : OneOf {
        /**
         * List of integers
         */
        lateinit var vecInt: Array<Int>
        /**
         * Long integer
         */
        var longInt: Long = 0L
    
        constructor(vecInt: Array<Int>, longInt: Long) {
            this.vecInt = vecInt
            this.longInt = longInt
        }
    
        /**
         * Write OptionOne to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, vecInt.size)
            for (vecIntElement in vecInt) {
                StreamUtil.writeInt(stream, vecIntElement)
            }
            StreamUtil.writeLong(stream, longInt)
        }
    
        /**
         * Get string representation of OptionOne
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("OptionOne { ")
            stringBuilder.append("vecInt: ")
            stringBuilder.append("[ ")
            var vecIntIndex = 0
            for (vecIntElement in vecInt) {
                if (vecIntIndex != 0) {
                    stringBuilder.append(", ")
                }
                stringBuilder.append(vecIntElement)
                vecIntIndex++
            }
            stringBuilder.append(" ]")
            stringBuilder.append(", ")
            stringBuilder.append("longInt: ")
            stringBuilder.append(longInt)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 0
    
            /**
             * Read OptionOne from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): OptionOne {
                var vecInt: Array<Int>
                vecInt = Array(StreamUtil.readInt(stream), {
                    var vecIntElement: Int
                    vecIntElement = StreamUtil.readInt(stream)
                    vecIntElement
                })
                var longInt: Long
                longInt = StreamUtil.readLong(stream)
                return OptionOne(vecInt, longInt)
            }
        }
    }

    /**
     * Second option
     */
    class OptionTwo : OneOf {
        /**
         * usize
         */
        var value: Int = 0
    
        constructor(value: Int) {
            this.value = value
        }
    
        /**
         * Write OptionTwo to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, value)
        }
    
        /**
         * Get string representation of OptionTwo
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("OptionTwo { ")
            stringBuilder.append("value: ")
            stringBuilder.append(value)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 1
    
            /**
             * Read OptionTwo from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): OptionTwo {
                var value: Int
                value = StreamUtil.readInt(stream)
                return OptionTwo(value)
            }
        }
    }
}

File: generated-code/python/example/TcpReadWrite/one_of.py
class OneOf:
    """Oneof example"""

    @staticmethod
    def read_from(stream):
        """Read OneOf from input stream
        """
        tag = stream.read_int()
        if tag == OptionOne.TAG:
            return OneOf.OptionOne.read_from(stream)
        if tag == OptionTwo.TAG:
            return OneOf.OptionTwo.read_from(stream)
        raise Exception("Unexpected tag value")

class OptionOne(OneOf):
    """First option"""

    TAG = 0

    def __init__(self, vec_int, long_int):
        self.vec_int = vec_int
        """List of integers"""
        self.long_int = long_int
        """Long integer"""

    @staticmethod
    def read_from(stream):
        """Read OptionOne from input stream
        """
        vec_int = []
        for _ in range(stream.read_int()):
            vec_int_element = stream.read_int()
            vec_int.append(vec_int_element)
        long_int = stream.read_long()
        return OptionOne(vec_int, long_int)
    
    def write_to(self, stream):
        """Write OptionOne to output stream
        """
        stream.write_int(self.TAG)
        stream.write_int(len(self.vec_int))
        for element in self.vec_int:
            stream.write_int(element)
        stream.write_long(self.long_int)
    
    def __repr__(self):
        return "OptionOne(" + \
            repr(self.vec_int) + \
            ", " + \
            repr(self.long_int) + \
            ")"

OneOf.OptionOne = OptionOne

class OptionTwo(OneOf):
    """Second option"""

    TAG = 1

    def __init__(self, value):
        self.value = value
        """usize"""

    @staticmethod
    def read_from(stream):
        """Read OptionTwo from input stream
        """
        value = stream.read_int()
        return OptionTwo(value)
    
    def write_to(self, stream):
        """Write OptionTwo to output stream
        """
        stream.write_int(self.TAG)
        stream.write_int(self.value)
    
    def __repr__(self):
        return "OptionTwo(" + \
            repr(self.value) + \
            ")"

OneOf.OptionTwo = OptionTwo